scrapers/bank_of_zambia.py

```diff
@@ -1,42 +1,87 @@
+import re
+
 import requests
 from playwright.sync_api import sync_playwright
 from utils import (
+    calculate_checksum,
     download_directory,
     get_browser,
     get_file_path,
     get_md5hash,
     get_standardized_filename,
     parse_arguments,
+    zip_pdfs,
 )
 
 BASE_URL = "https://www.boz.zm/"
 CIRCULAR_URL = f"{BASE_URL}circulars.htm"
 
 
-def bank_of_zambia_scraper():
+def get_link_id(link):
+    parsed_link = link[:20]
+    pattern = r"[^0-9]"
+    return re.sub(pattern, "", parsed_link)
+
+
+def bank_of_zambia_scraper(n=10):
     with sync_playwright() as pw:
         browser = get_browser(pw)
         page = browser.new_page()
         page.goto(CIRCULAR_URL)
 
         session = requests.Session()
 
-        pdf_link = page.locator("div.list-group > a").all()[0]
-        link = pdf_link.get_attribute("href")
+        pdf_links = page.locator("div.list-group > a").all()
+        count = min(n, len(pdf_links))
+
+        map_idx_to_link = {}
+        map_link_id_to_indices = {}
+
+        for i, pdf_link in enumerate(pdf_links[:count]):
+            link = pdf_link.get_attribute("href")
+            if not link or not link.endswith(".pdf"):
+                browser.close()
+                raise ValueError("Invalid file URL.")
+
+            link_id = get_link_id(link)
+
+            map_link_id_to_indices.setdefault(link_id, []).append(i)
+
+            map_idx_to_link[i] = link
+
+        file_dictionary = {}
+
+        for link_id in map_link_id_to_indices:
+            filepaths = []
+
+            for idx in map_link_id_to_indices[link_id]:
+                curr_link = map_idx_to_link[idx]
+
+                absolute_url = f"{BASE_URL}{curr_link}"
+
+                response = session.get(absolute_url)
+
+                with download_directory(__file__) as directory:
+                    filename = get_standardized_filename(__file__, curr_link)
+                    filepath = get_file_path(directory, filename)
+                    with open(filepath, "wb") as file:
+                        file.write(response.content)
+                        filepaths.append(filepath)
 
-        if not link or not link.endswith(".pdf"):
-            browser.close()
-            raise ValueError("Invalid file URL.")
+            md5hash = calculate_checksum(filepaths)
+            print("link_id", link_id, "md5hash", md5hash)
 
-        absolute_url = f"{BASE_URL}{link}"
+            if len(filepaths) > 1:
+                with download_directory(__file__) as directory:
+                    zip_name = get_standardized_filename(__file__, link_id, ".zip")
+                    zip_path = get_file_path(directory, zip_name)
+                    zip_pdfs(zip_path, filepaths)
+                file_dictionary[zip_name] = {"filepath": zip_path, "md5hash": md5hash}
+            else:
+                file_dictionary[filename] = {"filepath": filepath, "md5hash": md5hash}
```

Comment:
Few comments here:

Zipping the files will most likely be fairly standard across scrapers. Let's move the operation to `utils.py`
```
      # This should be in utils.py
      def some_function_name(filepaths, <whatever kwargs we need>):
            if len(filepaths) > 1:
                with download_directory(__file__) as directory:
                    # Let's get `zip_pdfs()` to get the standardized filename and filepath
                    # as well, and return the zipped filepath name
                    return zip_pdfs(__file__, link_id)
             return filepath
```

Then here we can

```suggestion
            output_filepath = some_function_name(**kwargs) if len(filepaths) > 1 else filepath
            filename = <get it some way....>
            file_dictionary[filename] = {"filepath": output_filepath, "md5hash": md5hash}
```